### PR TITLE
Feature/nt 0017

### PR DESCRIPTION
**Pull request by damian0o**

### CODE REVIEW

File: server/utils.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export const isString = (obj: any): obj is string =>
+  Object.prototype.toString.call(obj) === "[object String]";
```
Comment (mras2303) [inline server/utils.ts]: https://lodash.com/docs/#isString

File: server/utils.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+export const isObject = (obj: any): obj is Object =>
+  obj !== null && typeof obj === "object";
```
Comment (mras2303) [inline server/utils.ts]: https://lodash.com/docs/#isPlainObject

File: client/src/containers/ResetPassword/ResetPassword.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import React from "react";
```
Comment (mras2303) [inline client/src/containers/ResetPassword/ResetPassword.stories.tsx]: I think, this is not component, so stories for that is unnecessary

File: client/src/containers/ResetPassword/ResetPassword.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  mapPropsToValues: props => {
+    return {
+      email: props.email || ""
+    };
+  },
+  validationSchema: ResetPasswordSchema,
+  handleSubmit: (values, bag) => {
+    axios
+      .post("/api/users/password/reset", values)
+      .then(x => {
```
Comment (mras2303) [inline client/src/containers/ResetPassword/ResetPassword.tsx]: x? What does it mean "x"?

File: client/src/containers/ResetPassword/ResetPassword.tsx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    <LinksWrapper>
+      <TextLink to="/login">Already have an account? Sign In</TextLink>
+      <TextLink to="/register">Don&apos;t have an account? Sign Up</TextLink>
+    </LinksWrapper>
+    <MadeWithLove />
+  </Container>
+);
+
+export default connect(
+  () => {},
```
Comment (mras2303) [inline client/src/containers/ResetPassword/ResetPassword.tsx]: use null if you don't need mapStateToProps

File: client/src/containers/ResetPassword/index.tsx [hunk 6]
```diff
@@ -0,0 +1 @@
+export { default } from "./ResetPassword";
```
Comment (mras2303) [inline client/src/containers/ResetPassword/index.tsx]: I don't catch this notation

File: client/src/containers/ResetPassword/ResetPassword.stories.tsx (hunk 3 above)
Comment (damian0o) [reply client/src/containers/ResetPassword/ResetPassword.stories.tsx]: Not necessary but i belive it is nice to have it

File: server/utils.ts (hunk 1 above)
Comment (damian0o) [reply server/utils.ts]: This is a typescript typeguard notation.

See -> obj is string =>

File: server/utils.ts (hunk 2 above)
Comment (damian0o) [reply server/utils.ts]: This is a typeguard have some more feature then is plain object.
Do you mean replace whole thing with lodash or just a body ?

File: server/utils.ts (hunk 1 above)
Comment (mras2303) [reply server/utils.ts]: Yes, but lodash has built-in method to check it.

File: server/utils.ts (hunk 2 above)
Comment (mras2303) [reply server/utils.ts]: I think to start implement it in this project. This lib should be helpful for some array, string, object variables

File: client/src/containers/ResetPassword/ResetPassword.tsx [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import TextHelper from "../../components/TextHelper/TextHelper";
+import TextLink from "../../components/TextLink/TextLink";
+import TextField from "../../components/TextField/TextField";
+import Button from "../../components/Button/Button";
+import Container from "../../components/Container/Container";
+import { IconAvatarLock } from "../../components/AvatarIcon/AvatarIcon";
+import TextHeading from "../../components/TextHeading/TextHeading";
+import { addAlert, CanAddAlert } from "../../store/actions/alertActions";
+
+const StyledFormikForm = styled(Form)`
```
Comment (kamilmoskal) [inline client/src/containers/ResetPassword/ResetPassword.tsx]: We use it in every form so we can extract this to components

File: client/src/containers/ResetPassword/ResetPassword.tsx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      <TextLink to="/register">Don&apos;t have an account? Sign Up</TextLink>
+    </LinksWrapper>
+    <MadeWithLove />
+  </Container>
+);
+
+export default connect(
+  null,
+  { addAlert }
+)(ResetPassword);
```
Comment (kamilmoskal) [inline client/src/containers/ResetPassword/ResetPassword.tsx]: I believe making 2+ components in 1 file is not good practise, for better readability and maintenance better to keep them in different component files

File: client/src/containers/UpdatePassword/UpdatePassword.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+      />
+      {touched.password && errors.password && (
+        <TextHelper error component="div">
+          <p style={{ marginBottom: "5px" }}>- {errors.password}</p>
+        </TextHelper>
+      )}
+      {touched.passwordConfirm && errors.passwordConfirm && (
+        <TextHelper error component="div">
+          <p style={{ marginBottom: "5px" }}>- {errors.passwordConfirm}</p>
+        </TextHelper>
```
Comment (kamilmoskal) [inline client/src/containers/UpdatePassword/UpdatePassword.tsx]: I think we can in one condition handle displaying all errors

File: client/src/containers/UpdatePassword/UpdatePassword.tsx [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        <TextLink to="/register">Don&apos;t have an account? Sign Up</TextLink>
+      </LinksWrapper>
+      <MadeWithLove />
+    </Container>
+  );
+};
+
+export default connect(
+  null,
+  { addAlert }
```
Comment (kamilmoskal) [inline client/src/containers/UpdatePassword/UpdatePassword.tsx]: again too big file with 2 components

File: client/src/containers/UpdatePassword/UpdatePassword.tsx [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  password: Yup.string()
+    .min(6, "Password length must be at least 6 characters")
+    .required("Password is required"),
+  passwordConfirm: Yup.string()
+    .oneOf([Yup.ref("password"), null], "Passwords must match")
+    .min(6, "Password repeat length must be at least 6 characters")
+    .required("Password repeat is required")
+});
+
+const UpdatePasswordFormWrapper = withFormik<
```
Comment (kamilmoskal) [inline client/src/containers/UpdatePassword/UpdatePassword.tsx]: I wonder if it's better structure than I did SignInForm.tsx? I mean keeping validationSchema above component and functions logic in the top / Formik form in the bottom of component. It isn't better than seperate component and logic due to add withFormik after component

File: client/src/containers/UpdatePassword/index.tsx [hunk 12]
```diff
@@ -0,0 +1 @@
+export { default } from "./UpdatePassword";
```
Comment (kamilmoskal) [inline client/src/containers/UpdatePassword/index.tsx]: For what we need it?

File: server/controllers/users/users.ts [hunk 13]
```diff
@@ -1,236 +1,306 @@
...
+          if (emailExist) {
+            return res
+              .status(400)
+              .json({ errors: [{ msg: messages.EMAIL_ALREADY_EXISTS }] });
+          }
+          user.email = email;
+          if (!req.user.admin) {
+            user.activationKey = bcrypt.hashSync(email).replace(/\//g, "");
+            user.emailVerified = false;
+            sendVerificationEmail(user);
```
Comment (kamilmoskal) [inline server/controllers/users/users.ts]: It won't work, this function was extracted to another file

**Comment by kamilmoskal** [comment]
There is also problem that i mentioned in trello card with this task, That every random user can send reset password request to any email that he knows, we need some validation for that. And we should divide tasks like this to at least 2 smaller task for example for backend and frontend, because it isn't readable at all with all this 26 files, and it would be easier to correct and add changes.

File: server/utils.ts (hunk 1 above)
Comment (damian0o) [reply server/utils.ts]: This is not the same. This is type guard. lodash is missing obj is string part.

File: server/utils.ts (hunk 2 above)
Comment (damian0o) [reply server/utils.ts]: This function is not the same from technical point of view. This is typescript type guard
`obj is Object`

File: client/src/containers/ResetPassword/ResetPassword.tsx (hunk 4 above)
Comment (damian0o) [reply client/src/containers/ResetPassword/ResetPassword.tsx]: Change to decomposition

File: client/src/containers/ResetPassword/ResetPassword.tsx (hunk 5 above)
Comment (damian0o) [reply client/src/containers/ResetPassword/ResetPassword.tsx]: Replaced with null

File: client/src/containers/ResetPassword/index.tsx (hunk 6 above)
Comment (damian0o) [reply client/src/containers/ResetPassword/index.tsx]: Just compare importing components with or without this statement and index.tsx file

`import SignUp from "./containers/SignUp/SignUp";
import ResetPassword from "./containers/ResetPassword";`

File: client/src/containers/ResetPassword/ResetPassword.tsx (hunk 7 above)
Comment (damian0o) [reply client/src/containers/ResetPassword/ResetPassword.tsx]: I dont want to bother with this in this PR. But yes can be unify.

File: client/src/containers/UpdatePassword/index.tsx (hunk 12 above)
Comment (damian0o) [reply client/src/containers/UpdatePassword/index.tsx]: Just compare importing this component in App.tsx file

    import SignUp from "./containers/SignUp/SignUp";
    import ResetPassword from "./containers/ResetPassword";

**Comment by damian0o** [comment]
What validation can be introduced? If user forget his password he know only email.

I believe this is still very simple functionality and should not be merged separately since each part wont work on its own at least for end user.

File: client/src/containers/ResetPassword/ResetPassword.tsx (hunk 8 above)
Comment (damian0o) [reply client/src/containers/ResetPassword/ResetPassword.tsx]: Divided and throw away withFormik

File: client/src/containers/UpdatePassword/UpdatePassword.tsx (hunk 9 above)
Comment (damian0o) [reply client/src/containers/UpdatePassword/UpdatePassword.tsx]: Done

File: client/src/containers/UpdatePassword/UpdatePassword.tsx (hunk 10 above)
Comment (damian0o) [reply client/src/containers/UpdatePassword/UpdatePassword.tsx]: Splitted

File: client/src/containers/UpdatePassword/UpdatePassword.tsx (hunk 11 above)
Comment (damian0o) [reply client/src/containers/UpdatePassword/UpdatePassword.tsx]: Rewrite without withFormik. I believe looks better

File: server/controllers/users/users.ts (hunk 13 above)
Comment (damian0o) [reply server/controllers/users/users.ts]: fixed

**Pull request merged**